### ISSUE TITLE
Operation on home 'username' failed: Failed to execute operation: Invalid argument (4k sector issue)

### ISSUE THREAD

**Issue by ghost** [systemd/systemd#25792]
### systemd version the issue has been seen with

252.3-1

### Used distribution

archlinuxarm (asahi linux)

### Linux kernel version used

6.1.0-asahi-2-2-edge-ARCH

### CPU architectures issue was seen on

aarch64 (arm64)

### Component

systemd-homed

### Expected behaviour you didn't see

should create a user

### Unexpected behaviour you saw

Dec 20 12:35:24 unix systemd-homed[708]: tho2: changing state absent → creating
Dec 20 12:35:24 unix systemd-homework[36211]: Using automatic default storage of 'luks'.
Dec 20 12:35:24 unix systemd-homework[36211]: Formatting tool for compiled-in default file system btrfs not available, falling back to ext4 instead.
Dec 20 12:35:24 unix systemd-homework[36211]: Sizing home to 83% of available disk space, which is 119.2G.
Dec 20 12:35:24 unix systemd-homework[36211]: Allocating image file completed.
Dec 20 12:35:24 unix systemd-homework[36211]: Writing of partition table completed.
Dec 20 12:35:24 unix systemd-homework[36211]: Setting up loopback device /dev/loop0 completed.
Dec 20 12:35:25 unix systemd-homework[36211]: LUKS formatting completed.
Dec 20 12:35:26 unix systemd-homework[36211]: Detected attempt for concurrent LUKS2 metadata update. Aborting operation.
Dec 20 12:35:26 unix systemd-homework[36211]: Failed to set up LUKS password for slot 0: Invalid argument
Dec 20 12:35:26 unix systemd-homed[708]: Operation on tho2 failed: Invalid argument
Dec 20 12:35:26 unix systemd-homed[708]: block device /sys/devices/virtual/block/loop0 has been removed.

Operation on home tho2 failed: Failed to execute operation: Invalid argument

### Steps to reproduce the problem

1. sudo homectl create tho2
2. enter password
3. confirm password

**Comment by poettering** [comment]
this smells as if some crypto module or DM module or so is missing in your kernel build

**Comment by treeshateorcs** [comment]
@marcan ?

**Comment by marcan** [comment]
No, something is weird with the loop device that systemd creates. I see it configuring it like this:

    [pid  1618] ioctl(8, LOOP_CONFIGURE, {fd=5, block_size=512, info={lo_offset=0x100000, lo_number=0, lo_flags=LO_FLAGS_AUTOCLEAR|LO_FLAGS_DIRECT_IO, lo_file_name="", ...}}) = 0

Then it opens it and does a read:

    openat(AT_FDCWD, "/dev/loop1", O_RDONLY|O_DIRECT) = 9
    lseek(9, 0, SEEK_SET)       = 0
    read(9, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 512) = 512

And then on another fd, in RDWR mode:

    [pid  1618] openat(AT_FDCWD, "/dev/loop1", O_RDWR|O_DIRECT) = 10

It does a bunch of zeroing out:

    [pid  1618] lseek(10, 0, SEEK_SET)      = 0
    [pid  1618] ioctl(10, BLKZEROOUT, [0, 1048576]) = 0
    [pid  1618] lseek(10, 1048576, SEEK_SET) = 1048576
    [pid  1618] ioctl(10, BLKZEROOUT, [1048576, 1048576]) = 0
    [pid  1618] lseek(10, 2097152, SEEK_SET) = 2097152
    [...]
    [pid  1618] fsync(10) = 0

Then writes some randomness:

    [pid  1618] lseek(10, 8388608, SEEK_SET) = 8388608
    [pid  1618] read(5, "\"\203\243\277d\332\265\235\2I\220\356\200N\374\303`\177\275H`\355c\265\4\206\347\341I&\277\274"..., 1048576) = 1048576
    [pid  1618] write(10, "\"\203\243\277d\332\265\235\2I\220\356\200N\374\303`\177\275H`\355c\265\4\206\347\341I&\277\274"..., 1048576) = 1048576
    [...]
    [pid  1618] fsync(10) = 0

Then writes the header:

    [pid  1618] lseek(10, 0, SEEK_SET)      = 0
    [pid  1618] write(10, "LUKS\272\276\0\2\0\0\0\0\0@\0\0\0\0\0\0\0\0\0\2test2\0\0\0"..., 4096) = 4096
    [pid  1618] lseek(10, 4096, SEEK_SET)   = 4096
    [...]

Then tries to read it from the older fd, and that fails:

    [pid  1618] lseek(9, 0, SEEK_SET)       = 0
    [pid  1618] read(9, 0xaaab66848000, 512) = -1 EIO (Input/output error)

And it all goes downhill from there.

I can see a bug in systemd already: it's creating the image file with a 512 byte sector size, but these machines use SSDs with 4K sectors. I don't know if O_DIRECT on loop devices created in direct mode backed by a larger sector size block device is supposed to work at all, but even if it is, I'd expect broken corner cases and *terrible* performance due to RMW.

My bet is this will work if systemd learns to use the correct sector size for creating home loop images (or stops using direct backing mode).

**Comment by marcan** [comment]
Upon further thought, the above behavior is *exactly* what you would expect to happen due to the bad sector size. The first 512b read succeeds because it is on unallocated sectors, as the file is sparse (so there is no backing I/O to the device). The writes succeed because they are multiples of 4K. Then the next read fails because it actually hits the device as a 512b read and that's illegal with O_DIRECT on a 4K sector device (which is documented in the manpage).

So yeah, systemd needs to either switch the loop device to inherit the sector size from its backing device, or just switch to 4K unconditionally (which is probably a safe default that covers the vast majority of devices out there), or drop `LO_FLAGS_DIRECT_IO`.

**Comment by marcan** [comment]
Apparently there is an option for this already, and this works: `sudo homectl create --luks-sector-size=4096 foo`.

Obviously, that should be autodetected from the underlying FS's preferred IO size instead of blindly defaulting to 512.

**Comment by dtardon** [comment]
This is fixed by https://github.com/systemd/systemd/pull/26092, isn't it?

**Comment by poettering** [comment]
It might be. Not sure whether everything is addressed now. This is all complicated by the fact that disk sector sizes and O_DIRECT alignment requirements on an fs on top of that don't necessarily match up and there's no API to actually query the latter...

**Comment by poettering** [comment]
btw, would be grand if anyone who ran into this could give v253-rc1 a spin to verify this works correctly now.

**Comment by treeshateorcs** [comment]
@poettering when it's in the repos

**Comment by poettering** [comment]
@treeshateorcs i think at least fedora has builds of v253-rc1 in the repos already.

**Comment by treeshateorcs** [comment]
that doesn't help me, i'm running archlinuxarm

anyway, i took the PKGBUILD from version 252.4, lightly edited it and built  version 253~rc1

and it's the same error all over again

```
Jan 25 18:39:07 unix systemd-homed[504]: tho2: changing state absent → creating
Jan 25 18:39:07 unix systemd-homework[934]: Using automatic default storage of 'luks'.
Jan 25 18:39:07 unix systemd-homework[934]: Formatting tool for compiled-in default file system btrfs not available, falling back to ext4 instead.
Jan 25 18:39:07 unix systemd-homework[934]: Sizing home to 83% of available disk space, which is 54.3G.
Jan 25 18:39:07 unix systemd-homework[934]: Allocating image file completed.
Jan 25 18:39:07 unix systemd-homework[934]: Writing of partition table completed.
Jan 25 18:39:07 unix systemd-homework[934]: Setting up loopback device /dev/loop1 completed.
Jan 25 18:39:08 unix systemd-homework[934]: LUKS formatting completed.
Jan 25 18:39:09 unix systemd-homework[934]: Detected attempt for concurrent LUKS2 metadata update. Aborting operation.
Jan 25 18:39:09 unix systemd-homework[934]: Failed to set up LUKS password for slot 0: Invalid argument
Jan 25 18:39:09 unix systemd-homed[504]: Operation on tho2 failed: Invalid argument
Jan 25 18:39:10 unix systemd-homed[504]: block device /sys/devices/virtual/block/loop1 has been removed.
```

**Comment by solsticedhiver** [comment]
Just trying to use **systemd-homed** and **homectl** for the first time on *archlinux x86_64* with **systemd-253.1** and I got excatly the same error:

    # homectl activate testuser
    🔐 Please enter password for user testuser: **********              
    Operation on home testuser failed: Failed to execute operation: Invalid argument

**Comment by mstephenson6** [comment]
Still an issue in 253.4-1-arch (asahi). Using @marcan's workaround to create new users with homectl and LUKS

**Comment by topimiettinen** [comment]
By git bisection I meant #26092.

**Comment by topimiettinen** [comment]
Faster test case:
```
# dd if=/dev/zero of=/foo bs=10M count=1
# mke2fs /foo
# systemd-dissect /foo
```
Output should contain
```
 Sec. Size: 4096
```
and 512 for bad version.

**Comment by topimiettinen** [comment]
@treeshateorcs @solsticedhiver @mstephenson6 Could you try #28453 if it fixes your problem?

**Comment by topimiettinen** [comment]
With #28484 there's much more progress with `homectl create foo` but the operation times out on my system.
